### ISSUE TITLE
Undocumented implementation of SGLD

### ISSUE THREAD

**Issue by duburcqa** [huanzhang12/SA_PPO#2]
You provide an implementation of [SGLD](https://github.com/huanzhang12/SA_PPO/blob/7f5193e770bc4b31dd7c1ddc6a866b28ba816659/src/policy_gradients/steps.py#L586) that is substantially different from what you describe in the article. The main differences are:
- the scaling factor of the random noise is not consistent with the article. To my understanding, it should be `np.sqrt(2 * beta / step_eps)` instead of `np.sqrt(2 * step_eps * beta)`
- the scheduling of the temperature by a factor `1/(i+2)^2` looks arbitrary and not explained anywhere in the code or the article.
- the use of the sign of the gradient instead of its actual value. I guess this design choice is inspired from PGD (aka. multi-iteration FGSM), but it is not motivated anywhere. Your algorithm is should rather be called PGD-SGLD than plain SGLD.
- the initialization of the perturbation is not zero, which is critical because it avoids the saddle point issue completely. The critical advantage of doing so has been discussed [here](https://arxiv.org/pdf/2001.03994.pdf). So I think the SGLD gradient is not even relevant anymore. Sadly, you do not motivate the choice of the initialization either.

**Comment by huanzhang12** [comment]
@duburcqa Thanks for looking into our code carefully and I appreciate your questions.
1. It is `np.sqrt(2 * step_eps * beta)`, not `np.sqrt(2 * beta / step_eps)`. The noise is in proportional to step size. Some papers use an inverse temperature (1/beta) which might confused you.
2. We want the temperature to slowly go zero. You can use any other reasonable scheduling as well. I don't think this matters a lot.
3. It's not related FGSM. The projection is needed because we solve a constrained minimization problem (you may call it as projected SGLD). See https://papers.nips.cc/paper/2015/file/c0f168ce8900fa56e57789e2a2f2c9d0-Paper.pdf,  http://proceedings.mlr.press/v134/lamperski21a/lamperski21a.pdf
4. I know the paper you mentioned very well. The initialization can be seen as an added noise **at the first step of SGLD** (the first step has 0 gradient in our case so the only thing added is the noise). Additionally, in SGLD, noise is also added in **every step** to ensure better convergence. I am not sure what kind of motivation is needed here (this is also not the main focus of our paper).

**Comment by duburcqa** [comment]
Thank for answering by questions, but I have a few remarks:
1. For me it is not clear why and I did not find any info about it in the literature. I would be interesting to hear your interpretation of it.
2. OK I can understand that.  But if it really does not matter, it should be removed then. Otherwise it harms reproducibility of the article for no real advantage.
3. I disagree on this point in particular. For me the clipping that comes after is the projection. Taking the sign of the gradient weighted by the step size in the update is not. 
4. When I say "motivation", I mean at least comment in the code to justify doing something different from what is stated in the article. Because it is harming reproducibility, so IMHO it has to be at least documented somewhere. I can understand that the article is may be not the best place since it is not the main focus, but since the part about SGLD is already in appendix one or two extra lines is not really a problem. I understand your point and it makes sense to me, but it would have been nice to find it mention somewhere since in most implementation I found it was actually initialized to zero.

**Comment by huanzhang12** [comment]
Thank you for the comments, and I hope things will be more clear now:
1. See e.g., eq. (1) of the first paper I linked above, basically there is a `\sqrt{\eta}` (`\eta` is the `step_eps`) term before the noise, that's where the sqrt term comes from. As I mentioned, you might got confused with the `beta` (temperature) term, which was written as `1/beta` (inverse temperature) in the paper. Both notations are used in literature.
2. In my opinion, the detail of the scheduling does not matter when writing a paper (especially it is not a paper focusing on SGLD), it's basically like a hyper-parameter (like learning rate), which a practitioner should experiment and choose. It shouldn't be removed.
3. I guess it's just a name convention problem - all these first order methods look similar. It's basically gradient descent + projection + noise, and in my opinion it's a variant of Langevin dynamics. You can feel free to call it the way you want.
4. The implementation is not really different from the paper. As I mentioned, in our problem, the first step has a zero gradient, so the first step of SLGD is equivalent to adding noise directly as an initialization. The `(i+2)` factor in the scheduler also comes from this fact (the index starts from `2` since the first SGLD step is done as the initialization, and when i=0 it's the second SGLD iteration so we did i+2).

Thanks again for your comments and I am glad to have the discussion with you.
